### ISSUE TITLE
Incremental garbage collection

### ISSUE THREAD

**Issue by branlwyd** [divviup/janus#1467]
The current garbage collection implementation requires deleting all related expired reports, aggregation jobs, and collection jobs for a given task in a single transaction. This has a few problems:

* For large-scale tasks, deleting this much data in a single transaction will likely be too expensive.
* For tasks of any scale, if the GC process falls behind, we don't have an effective way to "catch up" and may fall into the same transaction size issue as with large-scale tasks.
* Strange dependencies could exist between entities which could delay GC in confusing ways. For example, if a very old & a recent client report happened to be grouped together into an aggregation job, neither of the reports could be GC'ed until the recent client report was ready for GC. This would lead the very old report to be kept around much longer than it otherwise would be.
* The SQL queries used to implement GC are very complex and likely expensive.

For this reason, an "incremental" GC is desirable. This GC should be able to delete arbitrary expired client reports, aggregation jobs, and collection jobs (& related artifacts) without dependencies on other types of entities being deleted in the same transaction.

Design sketch:

# Incremental GC in Janus

## Client reports

Client reports are stored in the `client_reports` table.

This design will make the following changes:
* Any `client_reports` older than `report_expiry_age` will now be considered eligible for GC, independent of GC-eligibility of any other entities in the system.
* The foreign-key relation of `report_aggregations.client_report_id` on `client_reports.id` will be dropped, to allow client report deletion to occur independently of aggregation job deletion.
* All client-report read & write methods will be updated to act as if `client_reports` rows which are eligible for GC do not exist. (e.g.: read methods will return nothing; write methods will fail with an appropriate row-does-not-exist error)

## Aggregation jobs

Aggregation jobs are stored in the `aggregation_jobs` and `report_aggregations` tables.

This design will make the following changes:
* Any aggregation jobs whose maximum of `client_timestamp_interval` is older than `report_expiry_age` will now be considered eligible for GC, independent of GC-eligibility of any other entities in the system. (Report aggregations are considered to be a part of a single related aggregation job; they will be considered eligible for GC if their related aggregation job is.)
* All aggregation-job read & write methods will be updated to act as if `aggregation_jobs`/`report_aggregations` rows eligible for GC do not exist.
* The start of the aggregation process reads client reports to retrieve report share information. A missing report will now fail that report's aggregation (with `ReportDropped`) rather than failing the entire aggregation job.
* GC'ed aggregation jobs must be handled the same as abandoned aggregation jobs, i.e. they must update appropriate `batches.outstanding_aggregation_jobs` rows & possibly close `batches` based on those updates. (Likely implementation strategy is to update the `aggregation_job_writer` to handle deletion.)

## Collection jobs

Collection jobs & related artifacts have the most complicated data model.

Used by the Leader only:
* `collection_jobs`: stores collection requests received from the Helper, as well as information used in the collection process (such as final aggregate shares).
* `outstanding_batches`: for fixed-size tasks, stores batches which have been created but not yet collected.
* `batches`: stores Leader-only state for batches (i.e. fixed-size batches, or time-interval minimal intervals)

Used by the Helper only:
* `aggregate_share_jobs`: stores aggregate share requests received from the Leader, as well as information used in the collection process (such as the final aggregate share).

Used by both Leader & Helper:
* `batch_aggregations`: stores state for batches, such as the incremental aggregate share aggregate share.

This design will make the following changes:
* `batches`, `batch_aggregations`, and `outstanding_batches` will be considered part of the same entity for purposes of GC, and will be considered eligible for GC once the maximum of the `client_timestamp_interval` of the `batches` row is older than `report_expiry_age`.
* `collection_jobs` and `aggregate_share_jobs` use the same rule to determine GC-eligiblity, but this rule is query type-specific.
  * For time-interval tasks, `collection_jobs` and `aggregate_share_jobs` are considered eligible for GC if the minimum of the collection interval is older than `report_expiry_age`. (The minimum is used instead of the maximum to ensure that collection jobs are not GC'ed after their underlying aggregation information from `batch_aggregations`.)
  * For fixed-size tasks, `collection_jobs` and `aggregate_share_jobs` are considered eligible for GC if the related batch is eligible for GC or has been GC'ed. [This choice induces a dependency between GC of different kinds of entities, but this is likely survivable -- thoughts on alternate approaches welcome.]
* All read & write methods for these entities will be updated to act as if GC-eligible rows do not exist.

## Updates to Garbage Collector logic

The Garbage Collector binary will be updated such that it finds & deletes GC-eligible rows using an iterative strategy: queries will be used to identify & delete a configurable number of rows per transaction. In this approach, different kinds of entities can be deleted concurrently (e.g. the transaction to GC client reports can run concurrently with the transaction to GC aggregation jobs). The GC binary will continuously run until it finds nothing to GC, at which point it will backoff exponentially.

## Other considerations

Task metrics, as used by `divviup-api`, are currently derived based on counting actually-existing `client_reports` & `report_aggregations` rows. This is already incorrect with our existing GC (if it were enabled), and will continue to be incorrect with incremental GC. We will need to fix things up, but this is left for a future design.

**Comment by branlwyd** [comment]
> Beyond making read and write methods ignore GC-eligible objects, we may need to change some of our more complicated queries to do the same at the database level.

Agreed -- my implementation plan was to do the GC-eligibility-filtering at the DB level where possible.


> Separately, we may want to set a cap on report expiration time

Definitely agreed; at a product level we will want to cap GC time, and we should consider if & how much we want to let subscribers modify it.

**Comment by tgeoghegan** [comment]
This plan seems sound to me. 

> Task metrics, as used by `divviup-api`, are currently derived based on counting actually-existing `client_reports` & `report_aggregations` rows. This is already incorrect with our existing GC (if it were enabled), and will continue to be incorrect with incremental GC. We will need to fix things up, but this is left for a future design.

This is covered at least in part by https://github.com/divviup/janus/issues/1460.

**Comment by branlwyd** [comment]
We will need to fix up `read_batch_size` in `datastore.rs`; currently it is implemented by counting `report_aggregations`, but this could be incorrect in the face of GC.

The likely fix is to augment `batches` (or `outstanding_batches`) to include this count directly. We'd need to ensure that the logic to GC aggregation jobs also decreases `max_size` by the number of incomplete aggregations in a GC'ed aggregation job; `min_size` does not need to be decremented similarly since this is a count of completed aggregations, and completed aggregations are recorded in collection tables (e.g. `batch_aggregations`).

**Comment by branlwyd** [comment]
The change suggested by the previous comment also needs to be made to `get_filled_outstanding_batch`.

**Comment by branlwyd** [comment]
I changed the design a bit during implementation: the Helper does not write `batches`, so `aggregate_share_jobs` can't use the related batch to determine GC-eligiblity. Instead, I augment the `aggregate_share_jobs` with a `client_timestamp_interval` and use that, instead.

(We'll now have `client_timestamp_interval` in `batches`, `batch_aggregations`, and `aggregate_share_jobs`; we really need a simplifying redesign of collection-related storage, but I'm going to leave it as tech debt for now, unfortunately.)

**Comment by branlwyd** [comment]
A corner case: in time-interval tasks, `batches` may have no reports. Thus, using the client timestamp interval is ineffective: depending on how the GC condition is interpreted, such `batches` will be either instantly GC'ed, or never GC'ed, both of which are unacceptable. (fixed-size batches must have at least one report as batches are explicitly created by an aggregation job, which associates some reports to the batch)

Since this only affects time-interval tasks, I am going to fix this up by using the batch interval for time-interval tasks only.

**Comment by branlwyd** [comment]
We need to be able to write empty `batch_aggregations`, with an empty `client_timestamp_interval`. Unfortunately, this isn't so easy to fix: we could use the interval from the `batch_identifier` for time-interval tasks, but for fixed-size tasks there is nothing to use -- there will be no related `batches` row if we are the Helper.

I think we need to start writing `batches` when we're the Helper. Since fixed-size batches will never be empty, we can use the `batches`' `client_timestamp_interval`.

I might also consider switching to just use a server-timestamp, decided on initial write, to drive data deletion. This "simple" design has become rather complex (at least partially due to technical debt in the collection process, but still).

**Comment by branlwyd** [comment]
Thinking about the above, a server-timestamp on write won't work: we need to delete some rows simultaneously (e.g. `batch_aggregations` for the same batch but different `ord`s) to be sure we avoid strange/bad behavior. A server-timestamp wouldn't work for this. (this is also another reason using `batch_aggregations`' `client_timestamp_interval` won't work -- different `batch_aggregations` will have different `client_timestamp_interval`s, so they won't necessarily be deleted in tandem)

**Comment by branlwyd** [comment]
This has been implemented (the final PRs were #1603 & arguably #1619).
